proposals/0000-replacing-cabal-custom-build.md

```diff
@@ -0,0 +1,1775 @@
+
+### Motivation: a simplistic first design
+
+To explain the design we have arrived at for fine-grained pre-build rules, let
+us first consider what a first draft design, which accommodates both the design
+of HLS as well as the existing `Custom` setup scripts, might look like:
+
+```haskell
+type TentativeRules env = env -> IO [TentativeRule]
+data TentativeRule = TentativeRule
+  { dependencies :: [Dependency]
+  , results :: [Result]
+  , action :: [ResolvedLocation]
+               -- ^ locations of __dependencies__ (determined by the build system)
+           -> [ResolvedLocation]
+              -- ^ locations for __results__ (chosen by the build system)
+           -> IO ()
+  }
+```
+
+That is, rules are specified by a function that takes in an environment
+(which in practice consists of information known to `Cabal` after configuring,
+e.g. `LocalBuildInfo`, `ComponentLocalBuildInfo`).
+This function returns an `IO` action that computes a list of rules. Each rule
+declares its inputs and outputs, and from this information arises the dependency
+structure between rules.  
+To run a rule, the build system must determine specific locations for all
+these inputs and outputs, and pass them to the `action` in order to
+execute the rule. For example, the build system would search for `blah/foo/bar.y`
+in the source directories of the project, and pass an absolute path to the
+file it found to the `action`.
+
+### Proposed design of rules
+
+There are several shortcomings with the above simplistic design:
+
+  1. it does not fit well with the proposed IPC interface for hooks:
+     namely, the build tool should be able to query the separate hooks
+     executable to obtain all the hooks that a package with `Hooks` `build-type`
+     provides. Doing so with the design proposed above would require a mechanism for
+     serialising and deserialising the `IO` actions that execute the rules, which in
+     practice would mean providing a DSL for `IO` actions that can be serialised,
+
+  2. it lacks information that would allow us to determine when the rules need
+     to be recomputed:
+
+       a. if the rules were computed by invoking `ghc -M` (or similar), we would
+          need to recompute them if the user adds a new file that would
+          have been found by that call to `ghc -M`.
+
+       b. if the `env` environment changed, we might or might not need to recompute
+          the rules. We should have a mechanism for rules to declare what part
+          of the environment they depend on, so that we don't need to pessimistically
+          rerun the computation anew each time.
+
+We propose to fix (1) by adding an extra layer of indirection: instead of a
+`Rule` directly storing an `IO` action, it stores a reference to an action.
+We can then separately query the external hooks executable with this reference
+in order to run the action.
+
+We fix (2) by adding `monitoredValue` and `monitoredDirs` fields to `Rule`.
```

Comment:
I'm unsure why we can't do what Shake does by pushing the responsibility for deciding whether to rerun into a rule, and then implementing early cutoff on top of that.

Reply:
The logic for deciding whether to re-run cannot live inside the rules themselves, because rules are not persistent across invocations. This means that there is no way to keep track of whether e.g. a certain value has changed inside a rule. It's only the tool that queries for the set of all rules that can compare the result of multiple invocations and see whether things have changed or not.

Reply:
Shake explicitly puts the persistence outside the rule. It's similar to your `monitoredValue`. You get a fingerprint value, which is stored fro you _by the build system_, and passed to the rule when it is run. The rule computes the new value of the fingerprint, and can terminate early and say that nothing changed if that is the case.

https://hackage.haskell.org/package/shake-0.19.7/docs/Development-Shake-Rule.html#g:1 explains